### PR TITLE
Serve API docs from Turing API server

### PR DESCRIPTION
**Pull request by romanwozniak**
### Context:
This PR introduces [swagger-ui-generator](https://github.com/gojek/turing/compare/main...romanwozniak:swagger_gen_dist?expand=1#diff-8deb06e38b91d01d9817aeca9816a99a753c59caab600482a1332d9cd1680d65) script together with necessary refactoring to make it possible to serve API docs from turing-api server.

### Changes:
* Added [swagger-ui-generator](https://github.com/romanwozniak/turing/tree/swagger_gen_dist/scripts/swagger-ui-generator) that downloads specified version of static `swagger-ui` distribution and configures it to be used with a local spec file
* Added [Makefile](https://github.com/gojek/turing/compare/main...romanwozniak:swagger_gen_dist?expand=1#diff-86b84087c9937b69f4fc51b0bce8b8b128d604af74850a2190abd84b604c5ccc) to simplify resolving and merging multiple openapi specs into a single file and downloading swagger ui to be used with that bundled spec file.
* Refactored `web.ServeReactApp` into `server.ServeSinglePageApplication` to re-use it for serving both Turing-UI and Swagger-UI
* Refactored `cmd/main.go` to define all the http routes of turing API with `gorilla/mux` (it was previously mix of `gorilla/mux` routers and `http` package routes)
* Updated helm chart to remove swagger-ui container from the deployment

### CODE REVIEW

File: api/go.mod [hunk 1]
```diff
@@ -30,7 +30,7 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/rs/cors v1.7.0
 	github.com/stretchr/testify v1.6.1
-	github.com/tidwall/gjson v1.6.1
+	github.com/tidwall/gjson v1.6.5
```
Comment (romanwozniak) [inline api/go.mod:33]: To fix Depandabot's security vulnerability alert:
Screenshot 2021-10-07 at 01 47 47

File: api/turing/api/ensemblers_api_it_test.go [hunk 2]
```diff
@@ -1,14 +1,18 @@
 // +build integration
 
-package api
+package api_test
```
Comment (ashwinath) [inline api/turing/api/ensemblers_api_it_test.go:3]: Why is this required?

File: scripts/swagger-ui-generator/swagger-ui-generator.sh [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    -h, --help               Display help
+    -s, --spec-file          A file containing the API definition
+    -o, --output             The output path for the generated Swagger UI files
+    -v, --version            The Swagger-UI version to use (default: DEFAULT_SWAGGER_UI_VERSION)"
+EOF
+}
+
+main() {
+  local version="$DEFAULT_SWAGGER_UI_VERSION"
+  local output=
```
Comment (ashwinath) [inline scripts/swagger-ui-generator/swagger-ui-generator.sh:28]: Does this work? This looks weird, normally I would do something like:

```
function get_x {
  echo "hello"
}

function get {
  local variable=$(get_x)
}
```

It's a lot of args you are parsing but I did not know you could do this with local variables.

**Review by ashwinath — APPROVED** [review]

File: api/api/Makefile [hunk 4]
```diff
@@ -1,11 +1,20 @@
-SHELL := /bin/bash
-VERSION_NUMBER=$(if $(VERSION),$(VERSION),$(shell ../../scripts/vertagen/vertagen.sh -f docker))
+ROOT_DIR:=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
```
[the remark this answers is not in the document]
Comment (ashwinath) [reply api/api/Makefile:1]: oh it's `make` specific. ok

File: scripts/swagger-ui-generator/swagger-ui-generator.sh (hunk 3 above)
Comment (romanwozniak) [reply scripts/swagger-ui-generator/swagger-ui-generator.sh:28]: I believe it's a common approach if you need to assign values for multiple variables in a single function. In this specific example `parse_command_line` is sort of a closure function (if I can say so), that is executed in scope of a parent function (`main`) and have access to its local variables.

Here is the same pattern in [helm/chart-releaser-action](https://github.com/helm/chart-releaser-action):
https://github.com/helm/chart-releaser-action/blob/6eec360873ca10c8e462c87ad50c38d1a4332638/cr.sh#L87

File: api/api/Makefile [hunk 5]
```diff
@@ -1,11 +1,20 @@
...
 
-.PHONY: build-image
-build-image:
-	@docker build . \
-		--tag $(if $(DOCKER_REGISTRY),$(DOCKER_REGISTRY)/,)turing/turing-swagger:${VERSION_NUMBER}
+.PHONY: bundle-openapi
+bundle-openapi:
+	@docker run --rm \
+		--volume ${ROOT_DIR}:/local \
+		 openapitools/openapi-generator-cli:v5.1.1 generate \
```
[the remark this answers is not in the document]
Comment (romanwozniak) [reply api/api/Makefile:7]: Oh, that's a real nitpick 😆. Will fix it in the next PR
